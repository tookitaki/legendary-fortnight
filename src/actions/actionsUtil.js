export default function actionCreator(type, ...argNames) {
    return function (...args) {
      const action = { type }
      argNames.forEach((_, index) => {
        action[argNames[index]] = args[index]
      })
      return action
    }
}