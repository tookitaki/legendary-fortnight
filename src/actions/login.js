import actionCreator from '../utils/action';

export const LOGIN_USER = 'LOGIN_USER';
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS';
export const LOGIN_USER_FAIL = 'LOGIN_USER_FAIL';
export const LOGOUT_USER = 'LOGOUT_USER';
export const CLEAR_USER_DATA = 'CLEAR_USER_DATA';

export const loginUser = actionCreator(LOGIN_USER, 'payload');
export const loginUserSuccess = actionCreator(LOGIN_USER_SUCCESS, 'payload');
export const loginUserFail = actionCreator(LOGIN_USER_FAIL, 'payload');
export const clearUserData = actionCreator(CLEAR_USER_DATA);

export const logoutUser = () => ({
  type: LOGOUT_USER,
  meta: {
    api: true
  },
  payload: {
    path: 'users/session',
    method: 'delete',
    onSuccess: [clearUserData],
    onError: [clearUserData]
  }
});
