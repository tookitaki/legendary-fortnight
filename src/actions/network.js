export const START_NETWORK = 'START_NETWORK';
export const END_NETWORK = 'END_NETWORK';
export const SET_NETWORK_ERROR = 'SET_NETWORK_ERROR';
export const CLEAR_NETWORK_ERROR = 'CLEAR_NETWORK_ERROR';

export const startNetwork = (actionType) => ({
  type: START_NETWORK,
  payload: {
    actionType
  }
});

export const endNetwork = (actionType) => ({
  type: END_NETWORK,
  payload: {
    actionType
  }
});

export const setNetworkError = (actionType, error) => ({
  type: SET_NETWORK_ERROR,
  payload: {
    actionType,
    error
  }
});

export const clearNetworkError = (actionType) => ({
  type: CLEAR_NETWORK_ERROR,
  payload: {
    actionType
  }
});
