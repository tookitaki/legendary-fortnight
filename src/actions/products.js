import actionCreator from '../utils/action';

export const FETCH_PRODUCTS = 'FETCH_PRODUCTS';
export const SET_PRODUCTS = 'SET_PRODUCTS';

export const setProducts = actionCreator(SET_PRODUCTS, 'payload');

export const fetchProducts = () => ({
  type: FETCH_PRODUCTS,
  meta: {
    api: true
  },
  payload: {
    path: 'products',
    onSuccess: [setProducts]
  }
});
