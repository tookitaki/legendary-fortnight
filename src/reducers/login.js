import produce from 'immer';

import {
  LOGIN_USER,
  LOGIN_USER_SUCCESS,
  LOGIN_USER_FAIL,
  LOGOUT_USER
} from '../actions/login';
import createReducer from '../utils/reducer';

export const initialState = { loading: false, token: null, error: null };

const loginUser = produce((draft, action) => {
  draft.loading = true;
  draft.error = null;
});

const loginUserSuccess = produce((draft, { payload: { token } }) => {
  draft.loading = false;
  draft.token = token;
  draft.error = null;
});

const loginUserFail = produce((draft, { payload: { error } }) => {
  draft.loading = false;
  draft.error = error;
});

const logoutUser = () => {
  return initialState;
};

const login = createReducer(initialState, {
  [LOGIN_USER]: loginUser,
  [LOGIN_USER_SUCCESS]: loginUserSuccess,
  [LOGIN_USER_FAIL]: loginUserFail,
  [LOGOUT_USER]: logoutUser
});

export default login;
