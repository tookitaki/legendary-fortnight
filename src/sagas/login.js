import { all, takeEvery, put, call } from 'redux-saga/effects'
import { auth } from '../services/auth';
import {
  LOGIN_USER,
  LOGOUT_USER,
  logoutUser,
  loginUserSuccess
} from '../actions/login'

export function* logoutUserSaga() {
  localStorage.removeItem('token');
}

export function* loginUserSaga(action) {
  try {
    const authResponse = yield call(auth, action.token);
    if (authResponse.status === 200 && authResponse.data &&
      authResponse.data.token) {

      localStorage.setItem('token', authResponse.data.token);
      yield put(loginUserSuccess({ token: authResponse.data.token }));
    } else {
      localStorage.clear();
      yield put(logoutUser());
    }
  } catch (error) {
    yield put(logoutUser());
  }
}

export default function* loginSaga() {
  yield all([
    takeEvery(LOGIN_USER, loginUserSaga),
    takeEvery(LOGOUT_USER, logoutUserSaga),
  ])
}