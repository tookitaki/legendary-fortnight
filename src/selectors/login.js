export const getLoginDetails = ({ login: { loading, token, error } }) => ({
  loading,
  token,
  error
});
