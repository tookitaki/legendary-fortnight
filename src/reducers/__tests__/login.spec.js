import { loginUser, loginUserSuccess, logoutUser } from '../../actions/login';
import login from '../login';

describe('Test for Login reducers', () => {
  const initialState = {
    isLoggedIn: false,
    loading: false,
    error: null
  };

  it('Should update log-in-user details', () => {
    const action = loginUser();
    const actual = login(initialState, action);
    expect(actual).toEqual({
      isLoggedIn: false,
      loading: true,
      error: null
    });
  });

  it('Should update log-in-user-success details', () => {
    const data = {
      token: 'test-token',
      userId: 1,
      name: 'tdss_user'
    };
    const action = loginUserSuccess({ data });
    const actual = login(initialState, action);
    expect(actual).toEqual({
      isLoggedIn: true,
      loading: false,
      error: null,
      auth: {
        token: 'test-token',
        userId: 1,
        name: 'tdss_user'
      }
    });
  });

  it('Should update logout details', () => {
    const action = logoutUser();
    const actual = login(initialState, action);
    expect(actual).toEqual(initialState);
  });
});
