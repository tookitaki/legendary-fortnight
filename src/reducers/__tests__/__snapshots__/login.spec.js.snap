// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Login reducers Should update log-in-user details 1`] = `
Object {
  "error": null,
  "loading": true,
  "token": null,
}
`;

exports[`Test for Login reducers Should update log-in-user-success details 1`] = `
Object {
  "error": null,
  "loading": false,
  "token": undefined,
}
`;

exports[`Test for Login reducers Should update logout details 1`] = `
Object {
  "error": null,
  "loading": false,
  "token": null,
}
`;

exports[`Test for Login reducers should have correct initial state 1`] = `
Object {
  "error": null,
  "loading": false,
  "token": null,
}
`;
